fix problems with widget visibility in QAP

dterrahe opened the pull request:
When the Quick Access Panel is shown, selected widgets are moved from their original module UI into the QAP structure. Their current visibility and sensitivity are stored and they are forcibly shown, but made insensitive if the module had hidden them to make them inaccessible. When the QAP is torn down, all the widgets are put back into their correct places and their original visibility and sensitivity are restored.

This causes problems when their visibility changes during their stay in the QAP. For example, add "denoise (profiled)" _mode_ and _patch size_ widgets to the QAP. Change the _mode_ from nlm to wavelets. The _patch size_ widget is hidden directly by the module gui code (even though the QAP would have wanted to still show it, but only make it insensitive). Now when moving away from the QAP, the original visibility of patch size is restored, even though it should still be invisible in wavelets mode.

This PR gets rid of the hidden->insensitive transformation in QAP and makes sure the sensitivity of the original widget and parent are always synced with their QAP versions, so that gui code that hides/unhides the widget (either directly, or by hiding/show(all)ing its parent) has its intended effect and the visibility of all widgets is still correct after moving away from the QAP. This also correctly disables shortcuts for hidden widgets.

@AlicVB I understand the intend of the choice to have all widgets in QAP always visible to avoid confusion when a user adds one and nothing shows up because it is currently invisible. So I tried hard to replicate this behavior. However, forcing the widget to always be visible makes it impossible to monitor current gui code; gtk_widget_show doesn't trigger a signal if the widget was already shown (but insensitive). An alternative could be to put an indicator in the header, similar to the red warning triangle some modules use.

dterrahe commented:
Since I'm here, might as well as actions for the first two tabs in the modulegroups. The others are customisable, so a little more complex.

dterrahe commented:
@AlicVB any thoughts?

AlicVB commented:
Sorry for the delay...
This is on my todo. I just lack some free time this days... maybe tomorrow...

dterrahe commented:
No worries! Just wanted to make sure you hadn't missed it :-)

AlicVB commented:
I have not tested the PR yet, but I understand the issue you fix here.

Now, I find your solution a little bit "radical", esp. because users might not understand why the widget they have set doesn't appear (as you have spotted in your introduction)...
*maybe* we can try one of the 2 solutions : 
1. some icon in the header to warn the user like you propose
2. a "replacement" text in case the widget is hidden ("patch size not available") : if it's hidden from the start, it's easy. The tricky part is to handle the case where the visibility is changed afterward. Maybe we can use gtk events...

My preference would go to 2, but I understand this is more work (if possible at all)

dterrahe commented:
TBH I don't think 2 would be that complicated. We just need to create one extra item for each widget, a label with its name and the explanation that it is currently hidden, add it to `d->mod_vbox_basic` as well (just before or just after the widget _box_, doesn't matter) and flip its visibility to make it always the opposite of `(widget invisible or box invisible)` in the same event handler that is now used.

My dislike for this approach is more that it is inconsistent with what we do _everywhere_ else in the gui, that it takes up extra space and that it is confusing if we have a whole list of labels that don't do anything.

I guess there would be an even less obtrusive option 3, to have a _tooltip_ for the header explaining the situation. My thinking is that this is really a one-time education issue. After the user understands what's going on, they presumably don't need to be reminded every single time how the program works, so an annoying gui feature might be too much. Just my 2c.

AlicVB commented:
yep, you may be right...
But I think we still need to inform the user of what happens, either with an icon + a tooltip or just a tooltip.

What would be great also is to add "actually hiden" or something similar to the list of widgets in the preferences (the right-click menu on the QAP icon or the "full" setting window. If an user doesn't understand why the added widget doesn't appear/has been removed, he may very well attempt to add it again...

dterrahe commented:
Added an add/remove menu to the header of each module in qap that just lists the widgets for that module. This is the subtree of the qap button pop-up menu.

In the menu, you can now add _and_ remove widgets at all levels _and_ whether a widget is currently included in the qap is indicated with a tick mark. If it is not currently visible, this is indicated with an "inconsistent" indicator, and in the tooltip.

![image](https://user-images.githubusercontent.com/1549490/188756069-4619a66d-da50-4fad-9f23-9dd835cd354f.png)

@Nilvus the color/weight of the letters in the menu items is currently the same, whether they are active or not. That is fine as long as we can see by the tick mark if it is active, but the "inconsistent" mark overwrites the tick mark. Would this be fixable via CSS? It would be nice to have "inactive" be somewhere between "active" and "insensitive".

AlicVB commented:
@dterrahe : just tested, and I like you last changes.
May I be picky and ask to go even further ?
- I would be great if the "invisible" is shown not only is submenus of each module, but also in the main list of the already added items. maybe just with a text, not sure it's worth the addition of the 3 state checkbox
- in the "full" setting window, it would also be great to show the "invisible" text in the main list, not only in addition menu
- in the "full" setting window, in the addition menu, there's a little bug for invisible *and* already added item : the tooltip says "(null) (currently not visible)" (but I'm not completely sure it's due to the changes here...

dterrahe commented:
Fixed 2 and 3 (and a memory leak).

I'm not sure if 1 is worth the extra hassle of dealing with the fact I don't have access to the action details at that point. Anything besides just a tooltip would become messy I feel and a confused user might not even look for the item in that list, or hover over the item when they do see it, so would never see the tooltip. Instead, they will go through the menu again to now _really_ add the widget and there see the "inconsistent" indicator and the tooltip.

Also, maybe we should educate people to add additional widgets using either the per-module menu I just added or directly from the full module by switching to mapping mode and ctrl-clicking the widget. In those situations the hidden-ness is more obvious...

TurboGit reviewed (approved):
Works for me. Nice improvements. I'll wait for @AlicVB for the final OK.

AlicVB reviewed (approved):
> Fixed 2 and 3 (and a memory leak).
> 
> I'm not sure if 1 is worth the extra hassle

I've just tested again your changes and I would say that you have implemented 1 (with a tooltip) and not 2... But I trust you when you say that it would be complex, considering it's really not crucial, let's go like that !

Thanks a lot for this !

dterrahe commented:
> you have implemented 1 (with a tooltip) and not 2

You are right of course.

TurboGit commented:
Thanks, merged now.

Nilvus commented:
I have few times actually but I didn't forget your ping @dterrahe. Will see that ASAP.
